src/Tools/ExternalAccess/LiveShare/Classification/RoslynSyntaxClassificationService.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Classification.Classifiers;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServices.LiveShare;
+using Microsoft.VisualStudio.LanguageServices.LiveShare.CustomProtocol;
+using LS = Microsoft.VisualStudio.LiveShare.LanguageServices;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LiveShare.Classification
+{
+    internal class RoslynSyntaxClassificationService : ISyntaxClassificationService
+    {
+        private readonly AbstractLspClientServiceFactory _roslynLspClientServiceFactory;
+        private readonly RemoteLanguageServiceWorkspace _remoteLanguageServiceWorkspace;
+        private readonly ISyntaxClassificationService _originalService;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IExperimentationService _experimentationService;
+        private readonly IThreadingContext _threadingContext;
+
+        public RoslynSyntaxClassificationService(AbstractLspClientServiceFactory roslynLspClientServiceFactory, RemoteLanguageServiceWorkspace remoteLanguageServiceWorkspace, ISyntaxClassificationService originalService,
+            ClassificationTypeMap classificationTypeMap, IExperimentationService experimentationService, IThreadingContext threadingContext)
+        {
+            _roslynLspClientServiceFactory = roslynLspClientServiceFactory;
+            _remoteLanguageServiceWorkspace = remoteLanguageServiceWorkspace;
+            _originalService = originalService;
+            _classificationTypeMap = classificationTypeMap;
+            _experimentationService = experimentationService;
+            _threadingContext = threadingContext;
+        }
+
+        public void AddLexicalClassifications(SourceText text, TextSpan textSpan, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
+        {
+            if (_experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SyntacticExp_Local))
+            {
+                using (new RequestLatencyTracker(SyntacticLspLogger.RequestType.LexicalClassifications))
+                {
+                    _originalService.AddLexicalClassifications(text, textSpan, result, cancellationToken);
```

Comment:
definitely concerned about this one :)   lexical classification is something we got lots of feedback on about even minor delays in cpu constrained scenarios.

Reply:
(or, rather, concerned about the code below that pushes it remotely).

Reply:
we will see whether people who accepts remote work care these as much as people use this locally. and we are not forcing anyone to do anything remote. if they see benefit of doing things remotely and fine with losing something they had locally, we will see how much degrading of experiences compared to local they are okay with it.

if they can't accept any differences, they can use keep using local with local repo as they are doing right now with full roslyn installed locally.

Reply:
> we will see whether people who accepts remote work care these as much as people use this locally. a

Out of curiosity, how do you intend to measure that?  From my experience at MS measuring satisfaction or how unhappy people are is *exceptionally* hard.  For example, one of the biggest learnings (and teaching) from the analytics training is that people aren't necessarily willing to tell you when they're unhappy.  There is a strong expectation on their part that teams will know when they've done something that has made things worse and that they will address it because the problem is so obvious.  

Does the team intend to heavily dogfood this for example to get an "at the fingertips" sense of how the quality feels in comparison.

> if they can't accept any differences, they can keep using local with local repo as they are doing right now with full roslyn installed locally.

WHy is this pitched as an either/or.  i.e. the user must either accept a degraded experience with remote work, or a good experience with local?  It's possible to have both.

Reply:
>
> Out of curiosity, how do you intend to measure that? From my experience at MS measuring satisfaction or how unhappy people are is exceptionally hard
Yep definitely hard.  There are people continuously dogfooding liveshare/nexus which will help give us feedback on this change (including me).  Additionally, we will collect the latency data and see if it even falls into an acceptable range.  I wouldn't say this will answer a definitive "this is good" but rather let us know if it's even worth considering.

> 
> WHy is this pitched as an either/or. i.e. the user must either accept a degraded experience with remote work, or a good experience with local? It's possible to have both.

Just a clarification, we're considering many options, one of which is this one (remote lsp), others include textmate, new ideas VS code is experimenting with (TBD), and of course a lightweight roslyn.  This just gives us data to better inform our decision :).   We do not have a great answer to the problem, so we're trying things first.